File: Simulation/Bronze/P3_2017-Milk_Measurement/measurement.py
#  Bucket sort approach
#  T: O(N), M: O(101 * 3 + 3) ≈ O(1)

with open("measurement.in", "r") as fin:
    # Read 1st line: number of measurements
    N = int(fin.readline().strip())
    bessie_timestamp = [0] * 101
    elsie_timestamp = [0] * 101
    mildred_timestamp = [0] * 101
    first_timestamp, last_timestamp = 101, 0
    # For the next N lines create bucket sorted arrays for each cow
    for i in range(N):
        timestamp, cow, change = fin.readline().split()
        timestamp, change = int(timestamp), int(change)

        if cow == "Bessie":
            bessie_timestamp[timestamp] += change
        elif cow == "Elsie":
            elsie_timestamp[timestamp] += change
        else:
            mildred_timestamp[timestamp] += change

        first_timestamp = min(first_timestamp, timestamp)
        last_timestamp = max(last_timestamp, timestamp)

display_changes = 0
lead_cows = set(['B', 'E', 'M'])
bessie_milk = elsie_milk = mildred_milk = 7
# For each milk production change
for i in range(first_timestamp, last_timestamp + 1):
    if (bessie_timestamp[i] == 0 and
    elsie_timestamp[i] == 0 and
    mildred_timestamp[i] == 0):
        continue

    # Re-calculate the milk production for each cow change
    bessie_milk += bessie_timestamp[i]
    elsie_milk += elsie_timestamp[i]
    mildred_milk += mildred_timestamp[i]

    max_milk = max(bessie_milk, elsie_milk, mildred_milk)
    curr_lead_cows = set()
    if bessie_milk == max_milk:
        curr_lead_cows.add('B')
    if elsie_milk == max_milk:
        curr_lead_cows.add('E')
    if mildred_milk == max_milk:
        curr_lead_cows.add('M')

    # If the lead cows have changed, increment the display changes
    if not curr_lead_cows == lead_cows:
        display_changes += 1
        lead_cows = curr_lead_cows

print(display_changes, file=open("measurement.out", "w"))

File: Simulation/Bronze/P3_2017-Milk_Measurement/measurement.cpp
#include <bits/stdc++.h>
#define N_MAX 101

using namespace std;

int main() {
    freopen("measurement.in", "r", stdin);
    // Read 1st line: number of measurements
    int n;
    cin >> n;
    int bessieTimestamp[N_MAX] = {0};
    int elsieTimestamp[N_MAX] = {0};
    int mildredTimestamp[N_MAX] = {0};
    int firsTimestamp = N_MAX, lastTimestamp = 0;
    // For the next N lines create bucket sorted arrays for each cow
    for (int i = 0; i < n; i++) {
        int timestamp, change;
        string cow;
        cin >> timestamp >> cow >> change;

        if (cow == "Bessie") bessieTimestamp[timestamp] += change;
        else if (cow == "Elsie") elsieTimestamp[timestamp] += change;
        else mildredTimestamp[timestamp] += change;

        firsTimestamp = min(firsTimestamp, timestamp);
        lastTimestamp = max(lastTimestamp, timestamp);
    }

    int displayChanges = 0;
    unordered_set<char> leadCows = {'B', 'E', 'M'};
    int bessieMilk = 7, elsieMilk = 7, mildredMilk = 7;
    // For each milk production change
    for (int i = firsTimestamp; i <= lastTimestamp; i++) {
        if (bessieTimestamp[i] == 0 &&
        elsieTimestamp[i] == 0 &&
        mildredTimestamp[i] == 0)
            continue;

        // Re-calculate the milk production for each cow change
        bessieMilk += bessieTimestamp[i];
        elsieMilk += elsieTimestamp[i];
        mildredMilk += mildredTimestamp[i];

        int maxMilk = max(bessieMilk, max(elsieMilk, mildredMilk));
        unordered_set<char> currLeadCows;
        if (bessieMilk == maxMilk) currLeadCows.insert('B');
        if (elsieMilk == maxMilk) currLeadCows.insert('E');
        if (mildredMilk == maxMilk) currLeadCows.insert('M');

        // If the lead cows have changed, increment the display changes
        if (currLeadCows != leadCows) {
            displayChanges++;
            leadCows = currLeadCows;
        }
    }

    freopen("measurement.out", "w", stdout);
    cout << displayChanges << endl;
    return 0;
}